extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java

```diff
@@ -14,22 +17,47 @@
 
 public class VertxCurrentContextFactory implements CurrentContextFactory {
 
-    public static final String LOCAL_KEY = "io.quarkus.vertx.cdi-current-context";
+    private static final String LOCAL_KEY_PREFIX = "io.quarkus.vertx.cdi-current-context";
+
+    private final List<String> keys;
+    private final List<String> unmodifiableKeys;
+
+    public VertxCurrentContextFactory() {
+        // There will be only a few mutative operations max
+        this.keys = new CopyOnWriteArrayList<>();
+        // We do not want to allocate a new object for each VertxCurrentContextFactory#keys() invocation
+        this.unmodifiableKeys = Collections.unmodifiableList(keys);
+    }
 
     @Override
     public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class<? extends Annotation> scope) {
-        return new VertxCurrentContext<>();
+        String key = LOCAL_KEY_PREFIX + scope.getName();
+        keys.add(key);
```

Comment:
I would probably be nice to have something like this in the `CurrentContextFactory` interface:

```java
/**
 * It is expected that this method is called at most once for each scope type.
 */
<T extends ContextState> CurrentContext<T> create(Class<? extends Annotation> scope);
```

Reply:
Yep, that makes sense. I will add it.

Reply:
Alternatively, I was thinking the `keys` could be a set (something like `ConcurrentHashMap.newKeySet(4)` seems reasonable).

Reply:
I believe that it will be better to throw an exception...

Reply:
That's fair, but also not exactly necessary. We currently call `CurrentContextFactory.create()` on exactly once, on exactly one place, and in near future, that may increase to 2, so formalizing the "at most once for each scope type" restriction should be good enough IMHO.

Reply:
I think that it's better to fail with a clean message if the contract is not fulfilled (which may happen because the `CurrentContextFactory` is part of the public API)...

Reply:
Yeah, agree.

However, if I look at how you define the contract

```java
@throws IllegalStateException If the implementation does not support multiple current contexts for the same scope
```

it seems to me that the thread-local implementation doesn't _support multiple current contexts for the same scope_, except it won't throw. I actually don't even know what would it mean to have multiple current contexts for one scope.

Something like this would likely be more precise:

```
This method may be called at most once for each scope type.
Implementations are permitted (but not required) to detect multiple invocations for the same scope type and throw an exception.
```

Reply:
> it seems to me that the thread-local implementation doesn't support multiple current contexts for the same scope, except it won't throw.

That depends if you map `CurrentContext` to the _"current context"_ from the spec. If you do then it can't be supported. If you don't then each invocation of `ThreadLocalCurrentContextFactory.create()` creates a new `ThreadLocalCurrentContext` with its own thread local variable. The truth is that by definition there can't be more than one active contexts for a normal scope but that does not necessarily cause a conflict here: contexts can be stored in the thread local but not active.

Anyway, it's true that this API is currently only used internally for the request context. And in the future, we'll probably add few more usages.

> This method may be called at most once for each scope type.
Implementations are permitted (but not required) to detect multiple invocations for the same scope type and throw an exception.

This is IMO not useful at all. If we forbid multiple invocations then an exception must be always thrown...

Reply:
Fair enough, I won't complain anymore :-)

Reply:
I take it as a reasonable discussion and not complaints ;-)